flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogSchema.scala

```diff
@@ -0,0 +1,166 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.catalog
+
+import java.util
+
+import org.apache.calcite.linq4j.tree.Expression
+import org.apache.calcite.schema._
+import org.apache.flink.table.api.{DatabaseNotExistException, TableNotExistException}
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.collection.JavaConverters._
+
+/**
+  * This class is responsible for connect external catalog to calcite catalog.
+  * In this way, it is possible to look-up and access tables in SQL queries
+  * without registering tables in advance.
+  * The databases in the external catalog registers as calcite sub-Schemas of current schema.
+  * The tables in a given database registers as calcite tables
+  * of the [[ExternalCatalogDatabaseSchema]].
+  *
+  * @param catalogIdentifier external catalog name
+  * @param catalog           external catalog
+  */
+class ExternalCatalogSchema(
+    catalogIdentifier: String,
+    catalog: ExternalCatalog) extends Schema {
+
+  private val LOG: Logger = LoggerFactory.getLogger(this.getClass)
+
+  /**
+    * Looks up database by the given sub-schema name in the external catalog,
+    * returns it Wrapped in a [[ExternalCatalogDatabaseSchema]] with the given database name.
+    *
+    * @param name Sub-schema name
+    * @return Sub-schema with a given name, or null
+    */
+  override def getSubSchema(name: String): Schema = {
+    try {
+      val db = catalog.getDatabase(name)
+      if (db != null) {
+        new ExternalCatalogDatabaseSchema(db.dbName, catalog)
+      } else {
+        null
+      }
+    } catch {
+      case e: DatabaseNotExistException =>
+        LOG.warn(s"database $name does not exist in externalCatalog $catalogIdentifier")
+        null
+    }
+  }
+
+  /**
+    * Lists the databases of the external catalog,
+    * returns the lists as the names of this schema's sub-schemas.
+    *
+    * @return names of this schema's child schemas
+    */
+  override def getSubSchemaNames: util.Set[String] = catalog.listDatabases().toSet.asJava
+
+  override def getTable(name: String): Table = null
+
+  override def isMutable: Boolean = true
+
+  override def getFunctions(name: String): util.Collection[Function] =
+    util.Collections.emptyList[Function]
+
+  override def getExpression(parentSchema: SchemaPlus, name: String): Expression =
+    Schemas.subSchemaExpression(parentSchema, name, getClass)
+
+  override def getFunctionNames: util.Set[String] = util.Collections.emptySet[String]
+
+  override def getTableNames: util.Set[String] = util.Collections.emptySet[String]
+
+  override def contentsHaveChangedSince(lastCheck: Long, now: Long): Boolean = false
```

Comment:
Right, my understanding of the parameter is that Calcite asks "Has the content changed?" and if we say "false" it will look into the cache. If we tell Calcite "true" (the content has changed) it will go against the external catalog. That's why I was thinking that `true` is the right return value. Can you check Calcite's behavior for this function?